lib/demangler.js

```diff
@@ -0,0 +1,173 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+"use strict";
+
+const
+    SymbolStore = require('./symbol-store').SymbolStore,
+    utils = require('../lib/utils');
+
+class Demangler {
+    /**
+     * 
+     * @param {string} demanglerExe 
+     * @param {BaseCompiler} compiler 
+     */
+    constructor(demanglerExe, compiler) {
+        this.demanglerExe = demanglerExe;
+        this.symbolstore = null;
+        this.othersymbols = new SymbolStore();
+        this.result = {};
+        this.input = [];
+        this.includeMetadata = true;
+        this.compiler = compiler;
+
+        this.labelDef = /^([.a-z_$][a-z0-9$_@.]*):/i;
+        this.jumpDef = /(j.+|b|bl|blx)\s+([.a-z_$][a-z0-9$_@.]*)/i;
+        this.callDef = /call[q]?\s+([.a-z_$][a-z0-9$_@.]*)/i;
+
+        // symbols in a mov or lea command starting with an underscore
+        this.movUnderscoreDef = /mov.*\s(_[a-z0-9$_@.]*)/i;
+        this.leaUnderscoreDef = /lea.*\s(_[a-z0-9$_@.]*)/i;
+        this.quadUnderscoreDef = /\.quad\s*(_[a-z0-9$_@.]*)/i;
+    }
+
+    addMatchToOtherSymbols(matches) {
+        if (!matches) return false;
+
+        const midx = matches.length - 1;
+        this.othersymbols.add(matches[midx], matches[midx]);
+
+        return true;
+    }
+
+    collectLabels() {
+        for (var j = 0; j < this.result.asm.length; ++j) {
+            const line = this.result.asm[j].text;
+
+            let matches = line.match(this.labelDef);
+            if (matches) {
+                const midx = matches.length - 1;
+                this.symbolstore.add(matches[midx], matches[midx]);
+            }
+
+            if (this.addMatchToOtherSymbols(line.match(this.jumpDef))) continue;
+            if (this.addMatchToOtherSymbols(line.match(this.callDef))) continue;
+            if (this.addMatchToOtherSymbols(line.match(this.movUnderscoreDef))) continue;
+            if (this.addMatchToOtherSymbols(line.match(this.leaUnderscoreDef))) continue;
+            if (this.addMatchToOtherSymbols(line.match(this.quadUnderscoreDef))) continue;
+        }
+
+        this.othersymbols.exclude(this.symbolstore);
+    }
+
+    getInput() {
+        this.input = [];
+        this.input = this.input.concat(this.symbolstore.listSymbols());
+        this.input = this.input.concat(this.othersymbols.listSymbols());
+
+        return this.input.join("\n");
+    }
+
+    /**
+     * 
```

Comment:
I could, but as it's being used currently, the doc is correct. But because the implementation in this class is empty, I removed the parameter... The actual implementation is in demangler-cpp, but it's called with the symbol parameter in this class